File: michangsheng/ItemSystem/items/alchemy/emulator.py
from typing import Iterator
from items.models.danyao import DanYao, Herbs
from items.loaders.danfang import DangFang
from items.alchemy.tables import tables


class Emulator:
    def __init__(self, target: DanYao, max_num: int = 9):
        """
        :param target: 需要模拟的对象
        :param max_num: 最大的草药限制
        """
        self.target = target
        self.max_num = max_num

    def generator(self, fuzzy: bool = False) -> Iterator:
        """生成所有丹方"""
        max_num = self.max_num
        for main1_num, main1_herb in tables.list_possible_main(
            key=self.target.main1, max_num=max_num, fuzzy=fuzzy
        ):
            possible_main2s = (
                tables.list_possible_main(
                    key=self.target.main2, max_num=max_num - main1_num, fuzzy=fuzzy
                )
                if self.target.main2
                else [(0, None)]
            )
            for main2_num, main2_herb in possible_main2s:
                possible_deputy1s = (
                    tables.list_possible_deputy(
                        self.target.deputy1,
                        max_num=max_num - main1_num - main2_num,
                        fuzzy=fuzzy,
                    )
                    if self.target.deputy1
                    else [(0, None)]
                )
                for deputy1_num, deputy1_herb in possible_deputy1s:
                    possible_deputy2s = (
                        tables.list_possible_deputy(
                            key=self.target.deputy2,
                            max_num=max_num - main1_num - main2_num - deputy1_num,
                            fuzzy=fuzzy,
                        )
                        if self.target.deputy2
                        else [(0, None)]
                    )
                    for deputy2_num, deputy2_herb in possible_deputy2s:
                        for intro_num, inteo_herb in tables.list_possible_intro(
                            key=self.target.intro,
                            max_num=max_num
                            - main1_num
                            - main2_num
                            - deputy1_num
                            - deputy2_num,
                            fuzzy=fuzzy,
                        ):
                            if self.check_neutralize(
                                inteo_herb,
                                main1_herb,
                                main2_herb,
                                deputy1_herb,
                                deputy2_herb,
                            ):
                                yield DangFang(
                                    id=0,
                                    ItemID=self.target.id,
                                    name=self.target.name,
                                    castTime=self.target.cost_time,
                                    value1=inteo_herb.id,
                                    num1=intro_num,
                                    value2=main1_herb.id,
                                    num2=main1_num,
                                    value3=main2_herb.id if main2_herb else 0,
                                    num3=main2_num,
                                    value4=deputy1_herb.id if deputy1_herb else 0,
                                    num4=deputy1_num,
                                    value5=deputy2_herb.id if deputy2_herb else 0,
                                    num5=deputy2_num,
                                )

    def check_neutralize(self, intro: Herbs, *args: Herbs) -> bool:
        """检测草药是否能中和"""
        flag = 0
        for herb in args:
            if herb is not None:
                if herb.intro.kind == 1:
                    flag -= 1
                elif herb.intro.kind == 2:
                    flag += 1

        return (flag > 0 and intro.intro.kind == 1) or (flag < 0 and intro.intro.kind == 2) or (flag == 0 and intro.intro.kind == 3)

    def __iter__(self):
        return self.generator()

File: michangsheng/FriendlyLianDan/liandan/Emulator.cs
using System.Collections.Generic;
using ItemSystem.Models;

namespace FriendlyLianDan
{
    public class Emulator
    {
        public DanYao target;
        public int maxNum;

        /*
        :param target: 需要模拟的对象
        :param maxNum: 最大的草药限制
        */
        public Emulator(DanYao target, int maxNum = 9)
        {
            this.target = target;
            this.maxNum = maxNum;
        }

        public IEnumerable<ItemSystem.Shims.DanFang> Generator(bool fuzzy = false)
        {
            if (target.danfang == null) {
                yield break;
            }

            foreach (Combination main1 in Tables.listPossibleMain(target.main1, fuzzy: fuzzy, maxNum: maxNum))
            {
                IEnumerable<Combination> possibleMain2s = target.main2 != null ? Tables.listPossibleMain(target.main2, fuzzy: fuzzy, maxNum: maxNum - main1.num) : new List<Combination> { new Combination(num: 0, herbs: null) };
                foreach (Combination main2 in possibleMain2s)
                {
                    IEnumerable<Combination> possibleDeputy1s = target.deputy1 != null ? Tables.listPossibleDeputy(target.deputy1, fuzzy: fuzzy, maxNum: maxNum - main1.num - main2.num) : new List<Combination> { new Combination(num: 0, herbs: null) };
                    foreach (Combination deputy1 in possibleDeputy1s)
                    {
                        IEnumerable<Combination> possibleDeputy2s = target.deputy2 != null ? Tables.listPossibleDeputy(target.deputy2, fuzzy: fuzzy, maxNum: maxNum - main1.num - main2.num - deputy1.num) : new List<Combination> { new Combination(num: 0, herbs: null) };
                        foreach (Combination deputy2 in possibleDeputy2s)
                        {
                            foreach (Combination intro in Tables.listPossibleIntro(target.intro, fuzzy: fuzzy, maxNum: maxNum - main1.num - main2.num - deputy1.num - deputy2.num))
                            {
                                if (ValidateNeutralize(intro.herbs.intro, main1.herbs, main2.herbs, deputy1.herbs, deputy2.herbs))
                                {
                                    var danfang = new ItemSystem.Shims.DanFang();
                                    danfang.name = target.name;
                                    danfang.ItemID = target.id;
                                    danfang.castTime = target.costTime;

                                    danfang.value1 = intro.herbs.id;
                                    danfang.num1 = intro.num;
                                    danfang.value2 = main1.herbs.id;
                                    danfang.num2 = main1.num;

                                    if (main2.herbs?.id != null)
                                    {
                                        danfang.value3 = main2.herbs.id;
                                    }
                                    danfang.num3 = main2.num;

                                    if (deputy1.herbs?.id != null)
                                    {
                                        danfang.value4 = deputy1.herbs.id;
                                    }
                                    danfang.num4 = deputy1.num;

                                    if (deputy2.herbs?.id != null)
                                    {
                                        danfang.value5 = deputy2.herbs.id;
                                    }
                                    danfang.num5 = deputy2.num;
                                    
                                    yield return danfang;
                                }
                            }
                        }
                    }
                }
            }
        }

        public static bool ValidateNeutralize(YaoXing intro, params Herbs[] args)
        {
            int flag = 0;
            foreach (Herbs herbs in args)
            {
                if (herbs != null)
                {
                    if (herbs.intro.kind == 1)
                    {
                        flag -= 1;
                    }
                    else if (herbs.intro.kind == 2)
                    {
                        flag += 1;
                    }
                }
            }
            return (flag > 0 && intro.kind == 1) || (flag < 0 && intro.kind == 2) || (flag == 0 && intro.kind == 3);
        }
    }
}